File: src/project_manager_flutter/lib/features/invoices/repository/invoice_repository.dart
import 'package:dio/dio.dart';
import '../../../core/constants/api_constants.dart';
import '../../../core/network/api_exceptions.dart';
import '../models/add_line_item_request.dart';
import '../models/generate_invoice_request.dart';
import '../models/invoice_detail.dart';
import '../models/invoice_line_item.dart';
import '../models/paginated_invoices.dart';
import '../models/update_invoice_status_request.dart';

class InvoiceRepository {
  final Dio _dio;

  InvoiceRepository(this._dio);

  Future<PaginatedInvoices> listInvoices({
    String? clientId,
    String? status,
    int page = 1,
    int pageSize = 20,
  }) async {
    try {
      final response = await _dio.get(
        ApiConstants.invoices,
        queryParameters: {
          'page': page,
          'pageSize': pageSize,
          if (clientId != null) 'clientId': clientId,
          if (status != null) 'status': status,
        },
      );
      return PaginatedInvoices.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<InvoiceDetail> getInvoice(String id) async {
    try {
      final response = await _dio.get(ApiConstants.invoiceById(id));
      return InvoiceDetail.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<Map<String, dynamic>> generateInvoice(GenerateInvoiceRequest request) async {
    try {
      final response = await _dio.post(ApiConstants.generateInvoice, data: request.toJson());
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<Map<String, dynamic>> updateInvoiceStatus(
      String id, UpdateInvoiceStatusRequest request) async {
    try {
      final response = await _dio.patch(
        ApiConstants.invoiceStatus(id),
        data: request.toJson(),
      );
      return response.data;
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }

  Future<InvoiceLineItem> addLineItem(String invoiceId, AddLineItemRequest request) async {
    try {
      final response = await _dio.post(
        ApiConstants.invoiceLineItems(invoiceId),
        data: request.toJson(),
      );
      return InvoiceLineItem.fromJson(response.data);
    } on DioException catch (e) {
      throw ApiException.fromResponse(e.response?.statusCode ?? 500, e.response?.data);
    }
  }
}

File: src/project-manager-web/src/features/invoices/repository/invoice-repository.ts
import { apiClient } from "@/core/api/client";
import { ApiEndpoints } from "@/core/api/constants";
import { ApiException } from "@/core/api/exceptions";
import type { Invoice, InvoiceDetail, GenerateInvoiceRequest, UpdateInvoiceStatusRequest, AddLineItemRequest } from "../models/types";
import type { PaginatedResponse } from "@/core/api/types";
import { AxiosError } from "axios";

function handleError(error: unknown): never {
  if (error instanceof AxiosError) {
    throw ApiException.fromResponse(error.response?.status ?? 500, error.response?.data);
  }
  throw error;
}

export const invoiceRepository = {
  async list(): Promise<Invoice[]> {
    try {
      const response = await apiClient.get<PaginatedResponse<Invoice>>(ApiEndpoints.invoices, {
        params: { pageSize: 100 },
      });
      return response.data.items;
    } catch (error) { handleError(error); }
  },

  async get(id: string): Promise<InvoiceDetail> {
    try {
      const response = await apiClient.get<InvoiceDetail>(ApiEndpoints.invoiceById(id));
      return response.data;
    } catch (error) { handleError(error); }
  },

  async generate(request: GenerateInvoiceRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.generateInvoice, request);
    } catch (error) { handleError(error); }
  },

  async updateStatus(id: string, request: UpdateInvoiceStatusRequest): Promise<void> {
    try {
      await apiClient.patch(ApiEndpoints.invoiceStatus(id), request);
    } catch (error) { handleError(error); }
  },

  async addLineItem(invoiceId: string, request: AddLineItemRequest): Promise<void> {
    try {
      await apiClient.post(ApiEndpoints.invoiceLineItems(invoiceId), request);
    } catch (error) { handleError(error); }
  },
};
